[ macOS ] 2 http/tests/media/FairPlay tests are broken.

jernoble opened the pull request:
#### 46b462205dd97be0f9eaa729085860c288436be1
<pre>
[ macOS ] 2 http/tests/media/FairPlay tests are broken.
<a href="https://bugs.webkit.org/show_bug.cgi?id=255294">https://bugs.webkit.org/show_bug.cgi?id=255294</a>
rdar://107891205

Reviewed by Ryosuke Niwa.

Fixing 4 related issues which keep these tests from passing consistentely:

1. When adding a MediaSource to a HTMLMediaElement via the element&apos;s .srcObject
   property, sometimes the MediaSource wrapper will be collected before the element
   has fully attached, and before the &quot;sourceopen&quot; event is fired. Mitigate this
   by returning true from virtualHasPendingActivity() in the time between when
   the MediaSource is set as srcObject and when it fires the &quot;sourceopen&quot; event.

2. HTMLMediaElement&apos;s own virtualHasPendingActivity() method will check whether
   the element is in the &quot;ended&quot; state. To do so, it calls m_player-&gt;duration(),
   which in the case of MediaSource, makes a call through a WeakPtr back up to
   the MediaSource to return _its_ duration. However, virtualHasPendingActivity()
   will be called on a background thread, and the WeakPtr accessor will throw
   an assertion. Rather than bounce down to MediaPlayer and back up to MediaSource,
   just query MediaSource directly.

   This will address the narrow issue of duration(), but the rest of the calls
   in virtualHasPendingActivity() may be non-thread-safe and further investigation
   and analysis is warranted.

3. Sometimes tests would fail with &quot;decode&quot; errors if ran back-to-back. One
   cause may be that CDM proxy objects in the GPU process are not consistently
   destroyed when their objects in the WebContent process are destroyed. Fix the
   RemoteLegacyCDMFactory to keep a WeakPtr map of CDM objects, and to track as
   those objects are created and destroyed, sending a message to the GPU process
   for its proxy to remove proxy objects.

   When the GPUConnectionToWebProcess closes, explicitly clear both the
   RemoteLegacyCDMFactory and RemoteCDMFactory objects, rather than wait for the
   connections refcount to go to zero.

   When destroying the CDMSessionAVContentKeySession, explicitly flush all
   attached SourceBufferPrivate objects, as well as removing the parser from
   the AVContentKeySession. Also explicitly expire the session before releasing
   it.

   This should cause system resources dedicated to protected playback to be more
   consistently freed after a test ends.

4. Add an explicit handler in the legacy tests which will fail the test (rather than
   time out) if the video throws an error. This will make the tests complete faster
   when video decoding fails, generating failed tests rather than time out results.

* LayoutTests/http/tests/media/fairplay/legacy-fairplay-hls.html:
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-v2.html:
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-v3.html:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::setAsSrcObject):
(WebCore::MediaSource::virtualHasPendingActivity const):
(WebCore::MediaSource::onReadyStateChange):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::setSrcObject):
(WebCore::HTMLMediaElement::durationMediaTime const):
(WebCore::HTMLMediaElement::detachMediaSource):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.mm:
(WebCore::CDMSessionAVContentKeySession::~CDMSessionAVContentKeySession):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::didClose):
* Source/WebKit/GPUProcess/media/RemoteCDMFactoryProxy.cpp:
(WebKit::RemoteCDMFactoryProxy::~RemoteCDMFactoryProxy):
(WebKit::RemoteCDMFactoryProxy::clear):
* Source/WebKit/GPUProcess/media/RemoteCDMFactoryProxy.h:
* Source/WebKit/GPUProcess/media/RemoteLegacyCDMFactoryProxy.cpp:
(WebKit::RemoteLegacyCDMFactoryProxy::~RemoteLegacyCDMFactoryProxy):
(WebKit::RemoteLegacyCDMFactoryProxy::clear):
* Source/WebKit/GPUProcess/media/RemoteLegacyCDMFactoryProxy.h:
* Source/WebKit/GPUProcess/media/RemoteLegacyCDMFactoryProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDMFactory.cpp:
(WebKit::RemoteLegacyCDMFactory::addSession):
(WebKit::RemoteLegacyCDMFactory::removeSession):
(WebKit::RemoteLegacyCDMFactory::didReceiveSessionMessage): Deleted.
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDMFactory.h:
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDMSession.cpp:
(WebKit::RemoteLegacyCDMSession::create):
(WebKit::RemoteLegacyCDMSession::~RemoteLegacyCDMSession):

Canonical link: <a href="https://commits.webkit.org/265252@main">https://commits.webkit.org/265252@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3315461f7ce8944bf3d87edd14cbfadb3e350db4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10280 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10518 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10784 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11927 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9894 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10292 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12512 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10468 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12849 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10438 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11170 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8661 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12313 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8477 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9310 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16589 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9576 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9461 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12710 "6 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9912 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8043 "3 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9071 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/9203 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2493 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13321 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9751 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -1055,6 +1065,7 @@ void MediaSource::onReadyStateChange(ReadyState oldState, ReadyState newState)
         buffer->readyStateChanged();
 
     if (isOpen()) {
+        m_sourceopenPending = false;
```
Are there failure cases where `m_sourceopenPending` should be set to `false` to allow the MediaSource wrapper to be collected (outside of when `HTMLMediaElement::detachMediaSource` is called), or will the `sourceopen` event always fire in response to setting `srcObject`?

jernoble replied:
`"sourceopen"` should always fire in response to setting `srcObject`.

jernoble committed:
[ macOS ] 2 http/tests/media/FairPlay tests are broken.
https://bugs.webkit.org/show_bug.cgi?id=255294
rdar://107891205

Reviewed by Ryosuke Niwa.

Fixing 4 related issues which keep these tests from passing consistentely:

1. When adding a MediaSource to a HTMLMediaElement via the element's .srcObject
   property, sometimes the MediaSource wrapper will be collected before the element
   has fully attached, and before the "sourceopen" event is fired. Mitigate this
   by returning true from virtualHasPendingActivity() in the time between when
   the MediaSource is set as srcObject and when it fires the "sourceopen" event.

2. HTMLMediaElement's own virtualHasPendingActivity() method will check whether
   the element is in the "ended" state. To do so, it calls m_player->duration(),
   which in the case of MediaSource, makes a call through a WeakPtr back up to
   the MediaSource to return _its_ duration. However, virtualHasPendingActivity()
   will be called on a background thread, and the WeakPtr accessor will throw
   an assertion. Rather than bounce down to MediaPlayer and back up to MediaSource,
   just query MediaSource directly.

   This will address the narrow issue of duration(), but the rest of the calls
   in virtualHasPendingActivity() may be non-thread-safe and further investigation
   and analysis is warranted.

3. Sometimes tests would fail with "decode" errors if ran back-to-back. One
   cause may be that CDM proxy objects in the GPU process are not consistently
   destroyed when their objects in the WebContent process are destroyed. Fix the
   RemoteLegacyCDMFactory to keep a WeakPtr map of CDM objects, and to track as
   those objects are created and destroyed, sending a message to the GPU process
   for its proxy to remove proxy objects.

   When the GPUConnectionToWebProcess closes, explicitly clear both the
   RemoteLegacyCDMFactory and RemoteCDMFactory objects, rather than wait for the
   connections refcount to go to zero.

   When destroying the CDMSessionAVContentKeySession, explicitly flush all
   attached SourceBufferPrivate objects, as well as removing the parser from
   the AVContentKeySession. Also explicitly expire the session before releasing
   it.

   This should cause system resources dedicated to protected playback to be more
   consistently freed after a test ends.

4. Add an explicit handler in the legacy tests which will fail the test (rather than
   time out) if the video throws an error. This will make the tests complete faster
   when video decoding fails, generating failed tests rather than time out results.

* LayoutTests/http/tests/media/fairplay/legacy-fairplay-hls.html:
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-v2.html:
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-v3.html:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::setAsSrcObject):
(WebCore::MediaSource::virtualHasPendingActivity const):
(WebCore::MediaSource::onReadyStateChange):
* Source/WebCore/Modules/mediasource/MediaSource.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::setSrcObject):
(WebCore::HTMLMediaElement::durationMediaTime const):
(WebCore::HTMLMediaElement::detachMediaSource):
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionAVContentKeySession.mm:
(WebCore::CDMSessionAVContentKeySession::~CDMSessionAVContentKeySession):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::didClose):
* Source/WebKit/GPUProcess/media/RemoteCDMFactoryProxy.cpp:
(WebKit::RemoteCDMFactoryProxy::~RemoteCDMFactoryProxy):
(WebKit::RemoteCDMFactoryProxy::clear):
* Source/WebKit/GPUProcess/media/RemoteCDMFactoryProxy.h:
* Source/WebKit/GPUProcess/media/RemoteLegacyCDMFactoryProxy.cpp:
(WebKit::RemoteLegacyCDMFactoryProxy::~RemoteLegacyCDMFactoryProxy):
(WebKit::RemoteLegacyCDMFactoryProxy::clear):
* Source/WebKit/GPUProcess/media/RemoteLegacyCDMFactoryProxy.h:
* Source/WebKit/GPUProcess/media/RemoteLegacyCDMFactoryProxy.messages.in:
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDMFactory.cpp:
(WebKit::RemoteLegacyCDMFactory::addSession):
(WebKit::RemoteLegacyCDMFactory::removeSession):
(WebKit::RemoteLegacyCDMFactory::didReceiveSessionMessage): Deleted.
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDMFactory.h:
* Source/WebKit/WebProcess/GPU/media/RemoteLegacyCDMSession.cpp:
(WebKit::RemoteLegacyCDMSession::create):
(WebKit::RemoteLegacyCDMSession::~RemoteLegacyCDMSession):

Canonical link: https://commits.webkit.org/265252@main
